loky/backend/reduction.py

```diff
@@ -170,29 +170,50 @@ def set_loky_pickler(loky_pickler=None):
-            _dispatch.update(_ReducerRegistry.dispatch_table)
-        else:
-            # Under Python 3 initialize the dispatch table with a copy of the
-            # default registry
-            _dispatch_table = copyreg.dispatch_table.copy()
-            _dispatch_table.update(_ReducerRegistry.dispatch_table)
+        def _set_dispatch_table(self, dispatch_table):
+            for ancestor_class in self._loky_pickler_cls.mro():
+                dt_attribute = getattr(ancestor_class, "dispatch_table", None)
+                if isinstance(dt_attribute, types.MemberDescriptorType):
+                    # Ancestor class (typically _pickle.Pickler) has a
+                    # member_descriptor for its "dispatch_table" attribute. Use
+                    # it to set the dispatch_table as a member instead of a
+                    # dynamic attribute in the __dict__ of the instance,
+                    # otherwise it will not be taken into account by the C
+                    # implementation of the dump method:
+                    # https://github.com/joblib/loky/pull/260
+                    dt_attribute.__set__(self, dispatch_table)
+                    return
+
+            # Fallback to regular setattr when inheriting from an alternative
+            # Pickler base class.
+            self.dispatch_table = dispatch_table
 
         def __init__(self, writer, reducers=None, protocol=HIGHEST_PROTOCOL):
             loky_pickler_cls.__init__(self, writer, protocol=protocol)
             if reducers is None:
                 reducers = {}
             if sys.version_info < (3,):
-                self.dispatch = self._dispatch.copy()
+                self.dispatch = loky_pickler_cls.dispatch.copy()
+                self._loky_dispatch.update(_ReducerRegistry.dispatch_table)
             else:
-                if getattr(self, "dispatch_table", None) is not None:
-                    self.dispatch_table.update(self._dispatch_table.copy())
+                if hasattr(self, "dispatch_table"):
+                    # Force a copy that we will update without mutating the
+                    # any class level defined dispatch_table.
+                    loky_dt = dict(self.dispatch_table)
                 else:
-                    self.dispatch_table = self._dispatch_table.copy()
+                    # Use standard reducers as bases
+                    loky_dt = copyreg.dispatch_table.copy()
+
+                # Register loky specific reducers
+                loky_dt.update(_ReducerRegistry.dispatch_table)
+
+                # Set the new dispatch table, taking care of the fact that we
+                # need to use the member_descriptor when we inherit from a
+                # subclass of the C implementation of the Pickler base class
+                # with an class level dispatch_table attribute.
+                self._set_dispatch_table(loky_dt)
 
+            # Register custom reducers
```

Comment:
`CustomizablePickler.register` still tries to modify `self.dispatch_table` directly, thus modifying `cloudpickle.Cloudpickler.dispatch_table` because of attribute shadowing. We should update `loky_dt` with the `reducers` dict before calling `self._set_dispatch_table`.

Reply:
Thanks for the explanations, that makes sense. Alternatively we could also setattr on top of __set__ and that should fix the problem. Let me give it a try.